Hw2/Task2/DiagonalMatrix.cs

```diff
@@ -0,0 +1,91 @@
+public class DiagonalMatrix
+{
+    private int[] _diagonal;
+    public int Size
+    {
+        get { return _diagonal.Length;}
+    }
+
+    public DiagonalMatrix(params int[] numbers) 
+    {
+        if(numbers == null) 
+        {
+            _diagonal = [];
+        }
+        else 
+        {
+            _diagonal = numbers;
+        }
+    }
+
+    public int this[int i, int j]
+    {
+        get 
+        {
+            if(i == j)
+            {
+                return _diagonal[i];
+            }
+            else if(i >= 0 && i < Size && j >= 0 && j < Size) 
+            {
+                return 0;
+            }
+            else
+            {
+                throw new IndexOutOfRangeException("Index out of range");
+            }
+        }
+        set 
+        { 
+            if(i == j)
+            {
+                _diagonal[i] = value;
+            }
+            else if(i < 0 || i >= Size || j < 0 || j >= Size) 
+            {
+                throw new IndexOutOfRangeException("Index out of range");
+            }
+        }
+    }
+
+    public int Track() 
+    {
+        return _diagonal.Sum();
+    }
+
+    public override bool Equals(object? obj)
+    {
+        if(obj is DiagonalMatrix dm) 
+        {
+            return Enumerable.SequenceEqual(_diagonal, dm._diagonal);
+        }
+        return false;
+    }
+
+    public override int GetHashCode()
+    {
```

Comment:
We were asked to override equals so i needed to override getHashCode as well. I dont think that there is reasonable implementation because  _diagonal is not immutable. I removed gethashcode.